contracts/appDefinitions/Battleship.sol

```diff
@@ -0,0 +1,204 @@
+pragma solidity 0.4.24;
+pragma experimental "ABIEncoderV2";
+
+import "openzeppelin-solidity/contracts/MerkleProof.sol";
+import "../lib/Transfer.sol";
+
+
+contract Battleship {
+  enum ActionType {
+    PLAY,
+    PLAY_AND_WIN
+  }
+
+  enum MoveType {
+    MISS_SQUARE,
+    HIT_SQUARE
+  }
+
+  struct Coord {
+    uint256 x;
+    uint256 y;
+  }
+
+  uint256 constant GAME_IN_PROGRESS = 0;
+  
+  // Empty square means not revealed yet
+  uint256 constant HIDDEN_SQUARE = 0;
+  uint256 constant SELECTED_SQUARE = 1; //Intermediate state between HIDDEN and HIT/MISS
+  uint256 constant MISS_SQUARE = 2;
+  uint256 constant HIT_SQUARE = 3;
+  uint256 constant TOTAL_SUNK_COUNT = 17; //Does not ensure correct placing of ships in any way
+
+  struct AppState {
+    address[2] players;
+    uint256 turnNum;
+    uint256 winner;
+    uint256[10][10] player1Board;
+    uint256[10][10] player2Board;
+    bytes32 player1MerkleRoot;
+    bytes32 player2MerkleRoot;
+    uint256 player1SunkCount; //How many squares were sunk. Total 17 squares sunk means all ships are sunk. ** Assuming placed correctly **
+    uint256 player2SunkCount;
+    uint256 prevMoveX;
+    uint256 prevMoveY;
+  }
+
+  struct Action {
+    /* Current move entries */
+    ActionType actionType;
+    uint256 currMoveX;
+    uint256 currMoveY;
+    /* Prev move entries */
+    uint256 prevMoveHitOrMiss; // MISS_SQUARE or HIT_SQUARE
+    uint256 prevMoveSalt;
+    bytes32[] prevMoveMerkleProof;
+  }
+
+  function isStateTerminal(AppState state)
+  public
+  pure
+  returns (bool)
+  {
+    return state.winner != GAME_IN_PROGRESS;
+  }
+
+  function getTurnTaker(AppState state)
+    public
+    pure
+    returns (uint256)
+  {
+    return state.turnNum % 2;
+  }
+
+  function applyAction(AppState state, Action action)
+  public
+  pure
+  returns (bytes)
+  {
+    require(state.winner==0, "Game has already been won");
+    AppState memory nextState;
+    if (action.actionType == ActionType.PLAY) {
+      nextState = playMove(state, action);
+    }
+    else if (action.actionType == ActionType.PLAY_AND_WIN) {
```

Comment:
This action should more accurately be called `WIN` since it does not invoke `playMove`